go/adbc/driver/bigquery/connection.go

```diff
@@ -0,0 +1,708 @@
+	scale, err := strconv.ParseInt(scaleString, 10, 32)
+	if err != nil {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse scale `%s` for field `%s`: %s", scaleString, name, err.Error()),
+		}
+	}
+	return int32(precision), int32(scale), nil
+}
+
+func (c *connectionImpl) getTableSchemaWithFilter(ctx context.Context, catalog *string, dbSchema *string, tableName string, columnName *string) (*arrow.Schema, error) {
+	if catalog == nil {
+		catalog = &c.catalog
+	}
+
+	if dbSchema == nil {
+		dbSchema = &c.dbSchema
+	}
+	sanitizedDbSchema, err := sanitizeDataset(*dbSchema)
+	if err != nil {
+		return nil, err
+	}
+
+	// sadly that query parameters cannot be used in place of table names
+	queryString := fmt.Sprintf("SELECT * FROM `%s`.`%s`.INFORMATION_SCHEMA.COLUMNS WHERE table_name = @tableName", *catalog, sanitizedDbSchema)
+	query := c.client.Query(queryString)
+	query.Parameters = []bigquery.QueryParameter{
+		{
+			Name:  "tableName",
+			Value: tableName,
+		},
+	}
+	reader, _, err := newRecordReader(ctx, query, c.Alloc)
+	if err != nil {
+		return nil, err
+	}
+
+	fields := make([]arrow.Field, 0)
+	var columns map[string]int = nil
+
+	for reader.Next() && ctx.Err() == nil {
+		rec := reader.Record()
+		if columns == nil {
+			columns = make(map[string]int)
+			for i, f := range reader.Schema().Fields() {
+				columns[strings.ToUpper(f.Name)] = i
+			}
+		}
+
+		numRows := int(rec.NumRows())
+		val, ok := columns["COLUMN_NAME"]
+		if !ok {
+			return nil, adbc.Error{
+				Code: adbc.StatusInternal,
+				Msg:  "Column `COLUMN_NAME` does not exist in response",
+			}
+		}
+		fieldName := rec.Column(val)
+
+		val, ok = columns["IS_NULLABLE"]
+		if !ok {
+			return nil, adbc.Error{
+				Code: adbc.StatusInternal,
+				Msg:  "Column `IS_NULLABLE` does not exist in response",
+			}
+		}
+		fieldNullable := rec.Column(val)
+
+		val, ok = columns["DATA_TYPE"]
+		if !ok {
+			return nil, adbc.Error{
+				Code: adbc.StatusInternal,
+				Msg:  "Column `DATA_TYPE` does not exist in response",
+			}
+		}
+		fieldType := rec.Column(val)
+
+		for i := 0; i < numRows; i++ {
+			metadata := make(map[string]string)
+			metadata["PRIMARY_KEY"] = ""
+			for j := 0; j < int(rec.NumCols()); j++ {
+				col := rec.Column(j)
+				columnName := strings.ToUpper(rec.ColumnName(j))
+				switch columnName {
+				case "COLUMN_NAME", "IS_NULLABLE":
+					continue
+				default:
+					metadata[columnName] = strings.Clone(col.ValueStr(i))
+				}
+			}
+
+			name := fieldName.ValueStr(i)
+			nullable := strings.ToUpper(fieldNullable.ValueStr(i))
+			dataType := fieldType.ValueStr(i)
+
+			field, err := buildSchemaField(name, dataType)
+			if err != nil {
+				return nil, err
+			}
+			field.Nullable = nullable == "YES"
+			field.Metadata = arrow.MetadataFrom(metadata)
+			fields = append(fields, field)
+		}
+		fieldName.Release()
+		fieldNullable.Release()
+		fieldType.Release()
+		rec.Release()
```

Comment:
since the record doesn't need to survive past each iteration of the loop, and you arent' calling `Retain` on the record, you don't need to explicitly release it. each call to `Next` will release the previous record for you automatically.